platform-sdk/swirlds-common/src/main/java/com/swirlds/common/io/streams/SerializableDataInputStream.java

```diff
@@ -586,4 +581,32 @@ private static <T extends SelfSerializable> T registryConstructor(final long cla
         }
         return rc;
     }
+
+    /**
+     * Reads a PBJ record from the stream.
+     *
+     * @param codec the codec to use to parse the record
+     * @param <T>   the type of the record
+     * @return the parsed record
+     * @throws IOException if an IO error occurs
+     */
+    public @NonNull <T extends Record> T readPbjRecord(@NonNull final Codec<T> codec) throws IOException {
+        final int size = readInt();
```

Comment:
This is a fixed int, 4 bytes. Why not a var int (and probably a protobuf tag, too), to make it a valid length-delimited protobuf field?

Reply:
Simple reason, I did not know how to do it 😄 . Are there any examples for this that I can look at?

Reply:
Reading and writing var ints is as simple as `readableSequentialData.readVarInt(false)` and `writableSequentialData.writeVarInt(value, false)`. Such a change does not make a lot of sense in itself, but if a protobuf tag is written / read before that, it would become a valid length-delimited protobuf message.

To work with tags, a `FieldDefinition` object is needed. Its name can be any, type is `FieldType.BYTES` or `FieldType.MESSAGE`, repeated=true, optional=true, oneof=false, number can be any. Then writing code will look like this:

```java
ProtoWriterTools.writeDelimited(writableSequentialData, fieldDefinition, codec.measureRecord(record), out -> codec.write(record, out));
```

It can be done in a separate PR.

Reply:
I created a ticket for it https://github.com/hashgraph/hedera-services/issues/15538